### PR TITLE
core: interrupt: update on client interface

### PR DESCRIPTION
**Pull request by ludovicbarre**
This patch series update client interface to:
- Take account type of interrupt in dt_get_irq() (SPI | PPI) and add
  interrupt resource in dt_node_info.
- Add an interface to allocate and add an interrupt handler.

### CODE REVIEW

File: core/kernel/dt.c [hunk 1]
```diff
@@ -4,6 +4,7 @@
  */
 
 #include <assert.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
```
Comment (jforissier) [inline core/kernel/dt.c:7]: This makes this file ARM-specific which is undesirable, especially now that we know that some people are porting OP-TEE to other architecture(s). Perhaps `dt_get_irq()` should remain declared in `core/include/kernel/dt.h` but its implementation should be moved under `core/arch/arm/kernel`?

File: core/include/kernel/dt.h [hunk 2]
```diff
@@ -106,7 +109,7 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname);
  * Returns the interrupt number if value >= 0
  * otherwise DT_INFO_INVALID_INTERRUPT
  */
-int dt_get_irq(void *fdt, int node);
+int dt_get_irq(const void *fdt, int node);
```
Comment (jforissier) [inline core/include/kernel/dt.h:112]: Unrelated change, should go in a separate commit.

File: core/kernel/dt.c [hunk 3]
```diff
@@ -52,7 +52,7 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
 	return prop;
 }
 
-int dt_get_irq(void *fdt, int node)
+int dt_get_irq(const void *fdt, int node)
```
Comment (jforissier) [inline core/kernel/dt.c:55]: Unrelated change, should go in a separate commit.

File: core/include/kernel/interrupt.h [hunk 4]
```diff
@@ -32,17 +32,24 @@ enum itr_return {
 	ITRR_HANDLED,
 };
 
+struct itr_handler;
+
+typedef enum itr_return (*itr_handler_t)(struct itr_handler *h);
```
Comment (jforissier) [inline core/include/kernel/interrupt.h:37]: We avoid typedefs as much as possible (coding rule, enforced by checkpatch). This one doesn't seem particularly useful since it is used only once locally, right?

File: core/include/kernel/interrupt.h [hunk 5]
```diff
@@ -32,17 +32,24 @@ enum itr_return {
...
 	void *data;
 	SLIST_ENTRY(itr_handler) link;
 };
 
 void itr_init(struct itr_chip *data);
 void itr_handle(size_t it);
 
+struct itr_handler *itr_alloc_add(size_t it, itr_handler_t handler,
+				  uint32_t flags, void *data);
+void itr_free(struct itr_handler *hdl);
```
Comment (jforissier) [inline core/include/kernel/interrupt.h:52]: `itr_free()` is used nowhere in this patch series so I'd rather add it when needed.

**Review by jforissier — COMMENTED** [review]
Hi @ludovicbarre,
Thanks for the patches, please see my comments below.

For commit "drivers: sp805_wdt: use itr_alloc_add()" you may add:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/kernel/dt.c (hunk 1 above)
Comment (ludovicbarre) [reply core/kernel/dt.c:7]: hi Jerome

like dt-bindings is specific by drivers, I could proposed you a second possibility:
- move the dt_get_irq in interrupt framework.
- Add a xlate function in itr_chip stucture, which allow to translate the specific properties of driver.
- this xlate function will be initialize by driver while init step
example: 
```
 void gic_init_base_addr(struct gic_data *gd, vaddr_t gicc_base __maybe_unused,
 			vaddr_t gicd_base)
 {
 	gd->gicc_base = gicc_base;
 	gd->gicd_base = gicd_base;
 	gd->max_it = probe_max_it(gicc_base, gicd_base);
 	gd->chip.ops = &gic_ops;
+#ifdef CFG_DT
+	gd->chip.xlate = gic_xlate;
+#endif
 }
```

-the dt_get_irq function find the interrupt property then call the dedicated xlate function of interrupt driver
example:
```
+#ifdef CFG_DT
+int dt_get_irq(const void *fdt, int node)
+{
+	const uint32_t *prop = NULL;
+	int len = 0;
+	int it_num = DT_INFO_INVALID_INTERRUPT;
+
+	if (!itr_chip || !itr_chip->xlate)
+		return it_num;
+
+	prop = fdt_getprop(fdt, node, "interrupts", &len);
+	if (!prop)
+		return it_num;
+
+	return itr_chip->xlate(prop, len);
+}
+#endif
```

what do you think about it ?
I will pull a fixup to show this proposal

File: core/include/kernel/interrupt.h (hunk 5 above)
Comment (ludovicbarre) [reply core/include/kernel/interrupt.h:52]: Like the interrupt framework provide the alloc Interface, I've prefer also provide the free Interface to the client.
This both functions integrated also  the list management Insert/Remove .

File: core/kernel/dt.c (hunk 1 above)
Comment (jforissier) [reply core/kernel/dt.c:7]: That would work, but is it useful? `xlate()` would be attached to `itr_chip` and there is only one instance of `itr_chip` in `core/kernel/interrupt.c`, so what does this bring us compared to a global function defined in the interrupt controller driver?

File: core/include/kernel/interrupt.h (hunk 5 above)
Comment (jforissier) [reply core/include/kernel/interrupt.h:52]: OK, no strong opinion on this, you may leave it if you wish but I have not changed my mind ;-)

File: core/kernel/dt.c (hunk 1 above)
Comment (ludovicbarre) [reply core/kernel/dt.c:7]: this allow to not duplicate the first part of code find the node "Interrupt", and translate the specific part in the driver.
If there are several architectures, the first part will be the same, but the second will be in own driver.

Comment (jforissier) [reply core/kernel/dt.c:7]: > this allow to not duplicate the first part of code find the node "Interrupt",

OK, but that's just one line of code isn't it? (`fdt_getprop()`). I am just trying to avoid over-engineering.

Comment (ludovicbarre) [reply core/kernel/dt.c:7]: Yes, that allow to split the responsibilities. the framework manage the property name "interrupt" and evolution (defined by dt-bindings or system device tree and which is not depend of cpu architecture), and avoid to define a dt_get_irq by architectures.
The drivers can focus only on translate parameters.
if you would, I could pull a fixup to show the small modification.

File: core/include/kernel/interrupt.h (hunk 5 above)
Comment (ludovicbarre) [reply core/include/kernel/interrupt.h:52]: no problem I understand the 2 POV:
-wait the need
-provide a symmetric interface alloc/free
thanks for your open mind ;-)

File: core/kernel/dt.c (hunk 1 above)
Comment (jforissier) [reply core/kernel/dt.c:7]: OK, let's see the code.

Comment (jforissier) [reply core/kernel/dt.c]: Rather than a fixup commit which removes this line I would prefer to see a clean series of patches, for instance in another PR. It is the best way to understand what is going on when reviewing.

**Comment by ludovicbarre** [comment]
hi Jerome

I push a fixup with the code, it's always more easy to speak with a peace of code

File: core/kernel/dt.c (hunk 1 above)
Comment (ludovicbarre) [reply core/kernel/dt.c]: ok, I will push a new PR. I come from the linux world. I don't know the procedures for optee yet. :-)

Comment (jforissier) [reply core/kernel/dt.c]: No worries -- in general we prefer fixups like you did (especially for minor fixes) but sometimes it's better to rework the patch series.

**Comment by ludovicbarre** [comment]
supersed by https://github.com/OP-TEE/optee_os/pull/4548
